awpy/analytics/map_control.py

```diff
@@ -0,0 +1,413 @@
+
+    Returns: TileNeighbors object with tile to neighbor mapping
+    """
+    tile_to_neighbors: TileNeighbors = defaultdict(set)
+
+    for tile_1, tile_2 in neighbor_pairs:
+        tile_to_neighbors[tile_1].add(tile_2)
+        tile_to_neighbors[tile_2].add(tile_1)
+
+    return tile_to_neighbors
+
+
+def calc_parsed_frame_map_control_values(
+    map_name: str,
+    current_player_data: FrameTeamMetadata,
+) -> FrameMapControlValues:
+    """Calculate tile map control values for each team given parsed frame.
+
+    Values are allocated to tiles depending on how many tiles are between it
+    and the source tile (aka tile distance). The smaller the tile distance,
+    the close the tile's value is to 1. The tile distance can get as high as
+    the max_depth argument, which is 10 as default. If the max_depth is kept
+    as default 10, this means the BFS search will stop once a tile
+    distance of 10 is reached.
+
+    Args:
+        map_name (str): Map used for find_closest_area and
+            relevant tile neighbor dictionary
+        current_player_data (FrameTeamMetadata): Object containing team metadata
+            (player positions, etc.). Expects extract_team_metadata output format
+
+    Returns: FrameMapControlValues object containing each team's map control values
+
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
+        raise ValueError(msg)
+
+    neighbors_dict = graph_to_tile_neighbors(list(NAV_GRAPHS[map_name].edges))
+
+    t_tiles = [
+        find_closest_area(map_name, i)["areaId"]
+        for i in current_player_data.t_metadata.alive_player_locations
+    ]
+    ct_tiles = [
+        find_closest_area(map_name, i)["areaId"]
+        for i in current_player_data.ct_metadata.alive_player_locations
+    ]
+
+    return _calc_frame_map_control_tile_values(
+        map_name, ct_tiles, t_tiles, neighbors_dict
+    )
+
+
+def calc_frame_map_control_values(
+    map_name: str,
+    frame: GameFrame,
+) -> FrameMapControlValues:
+    """Calculate tile map control values for each team given awpy frame object.
+
+    Values are allocated to tiles depending on how many tiles are between it
+    and the source tile (aka tile distance). The smaller the tile distance,
+    the close the tile's value is to 1. The tile distance can get as high as
+    the max_depth argument, which is 10 as default. If the max_depth is kept
+    as default 10, this means the BFS search will stop once a tile
+    distance of 10 is reached.
+
+    Args:
+        map_name (str): Map used for find_closest_area and
+            relevant tile neighbor dictionary
+        frame (GameFrame): Awpy frame object for map control calculations
+
+    Returns: FrameMapControlValues object containing each team's map control values
+
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
```

Comment:
this line also untested

Reply:
Added a simple sanity test for this line/function in next commit, thanks!